fprime-community/fpp
# C++ writer for component unit test code

Closes #284, closes #285, closes #287, closes #288, closes #289, closes #290.

- Add C++ writers for TesterBase and GTestBase classes
- Add C++ writers for Tester implementation templates, TesterHelpers.cpp, and TestMain.cpp
- Add command line option `-u` for generating unit test output for `fpp-to-cpp`, `fpp-depend`, and `fpp-filenames` (see [tools wiki](https://github.com/fprime-community/fpp/wiki/Tools))
- Add unit tests for test harness base classes and test harness implementation templates
- Revise `check-cpp` script for compilation testing for components
  - By default, it tests compilation with a few different combinations of F Prime guard values (same behavior as before)
  - When run with the `--all` flag, it tests compilation with all combinations of F Prime guard values, with the restriction that `FW_SERIALIZABLE_TO_STRING` and `FW_ARRAY_TO_STRING` always have the same value (before, `FW_AMPCS_COMPATIBLE` and `FW_CMD_CHECK_RESIDUAL` always remained at their default values of 0 and 1, respectively)
- Update unit tests for `fpp-depend` and `fpp-filenames`

Changes from Python autocoder

General 
- Prepend the component name to the generated file names, e.g. `[ComponentName]TesterBase.cpp` instead of `TesterBase.cpp`. This allows the generation of test harness classes for multiple components within the same directory.

TesterBase classes
- Reorganize class members (follows the same pattern as the reorganization of class members for component base classes, see #240)
- Change `typedef struct {...} [name];` to `struct [name] {...};`
- Move initialization of `m_param_[paramName]_valid` variable members from the `init` function to initializer lists in the constructor
- Fix issue where `[portName]_static` and dispatch function implementations for special ports are not generated when the corresponding special port constructs do not appear (e.g. a component contains command ports, but no commands)
- Change `sizeof([serializableType])` to `[serializableType]::SERIALIZED_SIZE` where appropriate in the `dispatchEvents` function
